src/ir/module-splitting.cpp

```diff
@@ -0,0 +1,405 @@
+  };
+  ModuleUtils::iterImportedMemories(primary, shareMemory);
+  ModuleUtils::iterDefinedMemories(primary, shareMemory);
+
+  auto shareTable = [&](Table* table) {
+    secondary.table.exists = true;
+    secondary.table.initial = table->initial;
+    secondary.table.max = table->max;
+    makeImportExport(*table, secondary.table, "table", ExternalKind::Table);
+  };
+  ModuleUtils::iterImportedTables(primary, shareTable);
+  ModuleUtils::iterDefinedTables(primary, shareTable);
+
+  auto shareGlobal = [&](Global* global) {
+    assert(primary.features.hasMutableGlobals() &&
+           "TODO: add wrapper functions for disallowed mutable globals");
+    auto secondaryGlobal = std::make_unique<Global>();
+    secondaryGlobal->type = global->type;
+    secondaryGlobal->mutable_ = global->mutable_;
+    secondaryGlobal->init =
+      global->init == nullptr
+        ? nullptr
+        : ExpressionManipulator::copy(global->init, secondary);
+    makeImportExport(*global, *secondaryGlobal, "global", ExternalKind::Global);
+    secondary.addGlobal(std::move(secondaryGlobal));
+  };
+  ModuleUtils::iterImportedGlobals(primary, shareGlobal);
+  ModuleUtils::iterDefinedGlobals(primary, shareGlobal);
+
+  auto shareEvent = [&](Event* event) {
+    auto secondaryEvent = std::make_unique<Event>();
+    secondaryEvent->attribute = event->attribute;
+    secondaryEvent->sig = event->sig;
+    makeImportExport(*event, *secondaryEvent, "event", ExternalKind::Event);
+    secondary.addEvent(std::move(secondaryEvent));
+  };
+  ModuleUtils::iterImportedEvents(primary, shareEvent);
+  ModuleUtils::iterDefinedEvents(primary, shareEvent);
+}
+
+template<class F> void forEachElement(Table& table, F f) {
+  for (auto& segment : table.segments) {
+    assert(segment.offset->is<Const>() &&
+           "TODO: handle non-constant segment offsets");
+    Index baseOffset = segment.offset->cast<Const>()->value.geti32();
+    for (size_t i = 0; i < segment.data.size(); ++i) {
+      f(Index(baseOffset + i), segment.data[i]);
+    }
+  }
+}
+
+// Return the first free table slot and the segment to which to append new
+// items, if it exists.
+Index getFirstFreeTableIndex(Table& table, Table::Segment** outSegment) {
+  Index firstFreeIndex = 0;
+  *outSegment = nullptr;
+  for (auto& segment : table.segments) {
+    assert(segment.offset->is<Const>() &&
+           "TODO: handle non-constant segment offsets");
+    Index segmentBase = segment.offset->cast<Const>()->value.geti32();
+    if (segmentBase + segment.data.size() > firstFreeIndex) {
+      firstFreeIndex = segmentBase + segment.data.size();
+      *outSegment = &segment;
+    }
+  }
+  return firstFreeIndex;
+}
+
+// Move the secondary functions out of the primary module and return a map from
+// their names to their (eventual) table indices.
+void moveFunctions(Module& primary,
+                   Module& secondary,
+                   const std::set<Name>& secondaryFuncs) {
```

Comment:
it may be slightly confusing that the set passed here is the ones to split out, but in the test it is the ones to keep. but maybe not worth changing.

Reply:
The public function matches the tests, so I think this is ok. Initially I was going to have the public function take a set of functions to split out, but I got feedback that any automatic analysis was more likely to identify the list of functions to keep.